src/ir/possible-contents.h

```diff
@@ -0,0 +1,577 @@
+  //       this and disallow ExactType etc., but PCV might get slower.
+  using Variant = std::variant<None, Literal, GlobalInfo, ExactType, Many>;
+  Variant value;
+
+public:
+  PossibleContents() : value(None()) {}
+  PossibleContents(const PossibleContents& other) : value(other.value) {}
+
+  // Most users will use one of the following static functions to construct a
+  // new instance:
+
+  static PossibleContents none() {
+    PossibleContents ret;
+    ret.value = None();
+    return ret;
+  }
+  static PossibleContents literal(Literal c) {
+    PossibleContents ret;
+    ret.value = c;
+    return ret;
+  }
+  static PossibleContents global(Name name, Type type) {
+    PossibleContents ret;
+    ret.value = GlobalInfo{name, type};
+    return ret;
+  }
+  static PossibleContents exactType(Type type) {
+    PossibleContents ret;
+    ret.value = ExactType(type);
+    return ret;
+  }
+  static PossibleContents many() {
+    PossibleContents ret;
+    ret.value = Many();
+    return ret;
+  }
+
+  PossibleContents& operator=(const PossibleContents& other) {
+    value = other.value;
+    return *this;
+  }
+
+  bool operator==(const PossibleContents& other) const {
+    return value == other.value;
+  }
+
+  bool operator!=(const PossibleContents& other) const {
+    return !(*this == other);
+  }
+
+  // Combine the information in a given PossibleContents to this one. The
+  // contents here will then include whatever content was possible in |other|.
+  void combine(const PossibleContents& other) {
+    // First handle the trivial cases of them being equal, or one of them is
+    // None or Many.
+    if (*this == other) {
+      return;
+    }
+    if (other.isNone()) {
+      return;
+    }
+    if (isNone()) {
+      value = other.value;
+      return;
+    }
+    if (isMany()) {
+      return;
+    }
+    if (other.isMany()) {
+      value = Many();
+      return;
+    }
+
+    auto type = getType();
+    auto otherType = other.getType();
+
+    if (!type.isRef() || !otherType.isRef()) {
+      // At least one is not a reference. We could in principle try to find
+      // ExactType here, say as the combination of two different constants, but
+      // as subtyping does not exist, Many is good enough anyhow, so do that.
+      value = Many();
+      return;
+    }
+
+    // Special handling for references from here.
+
+    // Nulls are always equal to each other, even if their types differ.
+    if (isNull() || other.isNull()) {
+      // If only one is a null, but the other's type is known exactly, then the
+      // combination is to add nullability (if the type is *not* known exactly,
+      // like for a global, then we cannot do anything useful here).
+      if (!isNull() && isTypeExact()) {
+        value = ExactType(Type(type.getHeapType(), Nullable));
+        return;
+      } else if (!other.isNull() && other.isTypeExact()) {
+        value = ExactType(Type(otherType.getHeapType(), Nullable));
+        return;
```

Comment:
This repetition could be avoided if this were a function operating on two `PossibleContent`s rather than a mutating method by canonicalizing to have the first of the two combined values be the null one, for instance. That would also let you combine the `value = *` lines with the `return` lines. Another way to combine those lines would be to implement e.g.

```c++
template<T> PossibleContents& operator=(T val) {
  *this = PossibleContents(val); return *this;
}
```

Reply:
I lean towards keeping this as a member function rather than static on two inputs, since the by-far common case is to keep merging in new contents to a particular location. It's possible clang is smart enough to avoid overhead of a static method + a copy, but I'd rather not risk it given how slow this operation already is...

Interesting idea to define `operator=(T val)` like that. I agree it could save a line or two. However, a worry I've had during development here is to make a mistake and do something like `PossibleContents(Type(x))` - that turns into an ExactType atm, but should probably be a ConeType when we add one later. So I think explicit constructors and assignments are safer than a template that might match more things than we expect.